mathesar/api/exceptions/mixins.py

```diff
@@ -0,0 +1,152 @@
+                if self.is_pretty(error):
+                    if 'field' not in error or error['field'] is None or str(error['field']) == 'None':
+                        error['field'] = error_type
+                    pretty.append(error)
+                else:
+                    pretty.extend(self.get_field_error_entries(errors[error_type], field))
+        if pretty:
+            return pretty
+        return []
+
+    def _run_validator(self, validator, field, message):
+        """
+        This method build on top of `_run_validator` method of the superclass
+        It provides the following additional features
+        1. Includes serializer if `required_context` is True similar to the behaviour of drf
+        """
+        try:
+            args = []
+            if getattr(validator, 'requires_context', False):
+                args.append(field)
+            validator(self.initial_data[field.field_name], *args)
+        except (DjangoValidationError, RestValidationError) as err:
+            err_message = err.detail[0] \
+                if hasattr(err, 'detail') else err.message
+            return err_message == message
+
+    @property
+    def errors(self):
+        """
+        This method build on top of `errors` property of the superclass to return a list instead of a dictionary
+        """
+        ugly_errors = super(FriendlyErrorMessagesMixin, self).errors
+        pretty_errors = self.build_pretty_errors(ugly_errors)
+        return ReturnList(pretty_errors, serializer=self)
+
+    @property
+    def field_map(self):
+        """
+        This method build on top of `field_map` property of the superclass
+        It provides the following additional features
+        1. Adds `ListSerializer` to `relation` field list
+        """
+        return {
+            'boolean': ['BooleanField', 'NullBooleanField'],
+            'string': ['CharField', 'EmailField', 'RegexField', 'SlugField',
+                       'URLField', 'UUIDField', 'FilePathField',
+                       'IPAddressField'],
+            'numeric': ['IntegerField', 'FloatField', 'DecimalField'],
+            'date': {'DateTimeField': self.DATETIME_FORMAT,
+                     'DateField': self.DATE_FORMAT,
+                     'TimeField': self.TIME_FORMAT,
+                     'DurationField': self.DURATION_FORMAT},
+            'choice': ['ChoiceField', 'MultipleChoiceField'],
+            'file': ['FileField', 'ImageField'],
+            'composite': ['ListField', 'DictField', 'JSONField'],
+            'relation': ['StringRelatedField', 'PrimaryKeyRelatedField',
+                         'HyperlinkedRelatedField', 'SlugRelatedField',
+                         'HyperlinkedIdentityField', 'ManyRelatedField', 'ListSerializer'],
+            'miscellaneous': ['ReadOnlyField', 'HiddenField', 'ModelField',
+                              'SerializerMethodField']
+        }
+
+    def get_field_kwargs(self, field, field_data):
```

Comment:
Again, can you reuse as much of the original method as possible? If the upstream package made improvements or fixes, it would be nice for us to reuse that as much as possible.

Reply:
Okay, thanks. Please do that and also create an issue in our repo to replace our code with the library's code ASAP.